internal/sources/file/file.go

```diff
@@ -0,0 +1,334 @@
+		tracker[fname] = fpos.pos
+	}
+	s.mapLock.RUnlock()
+
+	bts, err := json.Marshal(tracker)
+	if err != nil {
+		slog.Error(fmt.Sprintf("error marshalling offsets: %s", err))
+		return
+	}
+	bts = append(bts, '\n')
+	_, err = f.Write(bts)
+	if err != nil {
+		slog.Error(fmt.Sprintf("error writing to high watermark file: %s", s.highWatermarkFile))
+		return
+	}
+
+	slog.Debug(fmt.Sprintf("persisted offsets to high watermark file: %s", s.highWatermarkFile))
+
+	// rename the temporary file to the high watermark file
+	err = os.Rename(s.highWatermarkFile+".tmp", s.highWatermarkFile)
+	if err != nil {
+		slog.Error(fmt.Sprintf("error renaming high watermark file: %s", s.highWatermarkFile))
+		return
+	}
+}
+
+func (s *Watcher) loadOffsets() map[string]int {
+	ret := make(map[string]int)
+
+	// open the high watermark file
+	f, err := os.Open(s.highWatermarkFile)
+	if err != nil {
+		// if the file doesn't exist, return an empty map
+		slog.Info(fmt.Sprintf("cant open high watermark file: %s", err))
+		return ret
+	}
+	defer f.Close()
+
+	// read the offsets from the file
+	dec := json.NewDecoder(f)
+	err = dec.Decode(&ret)
+	if err != nil {
+		slog.Error(fmt.Sprintf("error decoding offsets: %s", err))
+		return ret
+	}
+	return ret
+}
+
+func (s *Watcher) commitLoop(ctx context.Context) error {
+	// Wait until we have loaded the initial offsets
+	select {
+	case <-s.loaded:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			s.persistOffsets()
+			return ctx.Err()
+		case <-s.commitTicker.C:
+			s.persistOffsets()
+		}
+	}
+}
+
+func (s *Watcher) recvLoop(ctx context.Context) error {
+
+	var callback = func(startOffset int) func(ctx context.Context, b []byte, fname string, pos int) error {
+		return func(ctx context.Context, b []byte, fname string, pos int) error {
+			event := kawa.Message[types.Event]{
+				Value: types.Event{
+					// TODO: how do we parse eventTime from the file?
+					EventTime:  time.Now(),
+					SourceType: "file",
+					RawLog:     b,
+				},
+			}
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case s.msgC <- msgErr[types.Event]{
+				msg: event,
+				ack: func() {
+					s.savePosition(fname, startOffset+pos)
+				},
+				err: nil,
+			}:
+			}
+			return nil
+		}
+	}
+
+	// load the offsets from the high watermark file
+	offsets := s.loadOffsets()
+	firstRun := true
+
+	for {
+		// list the files in the directory given by s.path
+		files, err := os.ReadDir(s.path)
+		if err != nil {
+			return fmt.Errorf("watcher: %w", err)
+		}
+
+		for _, file := range files {
+			if file.IsDir() {
+				continue
+			}
+			// resolve the full path of the file
+			fname := filepath.Join(s.path, file.Name())
+			if s.extension != "" && !strings.HasSuffix(fname, s.extension) {
+				slog.Debug(fmt.Sprintf("skipping file without given extension (%s): %s", s.extension, fname))
+				continue
+			}
+			s.mapLock.Lock()
+			shouldOpen := false
+
+			if cur, ok := s.fmap[fname]; ok {
```

Comment:
Assuming a scenario like this: The logging app is writing to `log.txt` and then that file is rotated to `log.txt.1` and the app creates a new `log.txt` and starts writing to it.

Does this handle that scenario correctly. I think it might have issues with the file name lookup and might start to re-process `log.txt.1` since it is now a different file name and not in the map.

It looks like it will handle `log.txt` being new correctly as long as `!os.SameFile` returns false, it will start reading from the beginning again.

Though I guess the way to fix this is to make the config extension value `log.txt` so that it only reads the newest file and not the others in the directory. If log rotation is setup in this manner.

Reply:
I think it would make sense to enforce the extension setting (or later on a set of extensions) to handle this case.

Most of the time the log files will be compressed and get a new extension.

Still though, it would be nice to handle that case with a little more grace.  I'll think a bit more about what we can do here.